Handle the list of Pull Requests

opudrovs opened the pull request:
Closes https://github.com/weaveworks/tf-controller/issues/662
Closes https://github.com/weaveworks/tf-controller/issues/586

- Added the `createTerraform`, `createSource`, `createLabels`, and `createObjectName` methods.

- Added the `listTerraformObjects` method.

- Added the `deleteTerraform` and `deleteSource` methods.

- Reworked branch-based planner label keys and values.

- Added the `branchPollingInterval` flag.

- Added Terraform object and source reconciliation logic.

- Added the `reconcileTerraform` and `reconcileObjects` functions for Terraform object and source reconciliation.

- Added tests for creating, updating, and deleting objects during polling

- Added a new test target to run tests in the `internal` directory.

- Added running internal tests to the `test` workflow.

squaremo commented on internal/server/polling/server.go:
```diff
@@ -97,9 +99,86 @@ func (s *Server) poll(ctx context.Context, resource types.NamespacedName, secret
+	}
+
+	// List Terraform objects, created by BBP.
+	tfList, err := s.listTerraformObjects(ctx, tf, annotations)
```
If you used a label for this purpose, you'd be able to use label selectors (the API server might even help out with indexing, and the controller-runtime cache certainly will).

squaremo replied:
(You could consider using onwer references for this, too; but I understand if you want to restrict the scope of this PR).

opudrovs replied:
@yitsushi is there a reason you want to use annotations instead of labels for defining which objects are managed by BBP?

yitsushi replied:
Nothing particular.

opudrovs replied:
OK, then I can switch to using labels.

squaremo commented on internal/server/polling/server.go:
```diff
@@ -97,9 +99,86 @@ func (s *Server) poll(ctx context.Context, resource types.NamespacedName, secret
+				continue
+			}
+
+			if pr.HeadBranch == branchSource.Spec.Reference.Branch {
```
Isn't there an annotation for the PR being tracked? Are there any ways the test in this line could succeed when you don't want it to?

opudrovs replied:
@squaremo do you mean to use PR ID instead of branch name here? Can do it, thanks.

squaremo reviewed (commented):
Just comments for considering, since this is a draft.

squaremo commented on internal/server/polling/terraform.go:
```diff
@@ -38,6 +61,114 @@ func (s *Server) getSource(ctx context.Context, tf *v1alpha2.Terraform) (*source
+		return fmt.Errorf("unable to create Source for %s: %w", msg, err)
+	}
+
+	tf.Spec.SourceRef.Name = source.ObjectMeta.Name
```
This relies on the (safe) assumption that the copy of the source must be in the same namespace as the original source. A comment explaining why this assumption is safe would help people read the code -- or you could just defensively assign the namespace, so people don't even have to pause and think about it.

opudrovs replied:
What could be a reason not to put branch-based sources to the same namespace as the original source?

squaremo commented on internal/server/polling/server.go:
```diff
@@ -97,9 +99,75 @@ func (s *Server) poll(ctx context.Context, resource types.NamespacedName, secret
+	// Create a list of objects to delete.
+	oldTFs := []*infrav1.Terraform{}
+
+	for _, branchTF := range tfList {
```
It's not crucial here, but this and the following loop are quadratic -- they both, worst case, look through all the PRs x all the Terraform objects. Can we improve on that? (I am happy for this to be a follow up, since I think this is _correct_ as it stands).

opudrovs replied:
How exactly would you like to improve it?

yitsushi replied:
I think as a first step, we can merge the two nested loops (delete tfs and create tfs).

For example:
```
// Start with PRs, they are the source of truth.
loop _, pr := prs {

  // Assume we have to create it.
  shouldCreate := true
  
  // Worst case, we iterated through all of them, that means we have to create one.
  loop idx, tf := tfList as tf {
    if pr belongs to tfList {
    
      // We don't care, there is a TF object for this PR, nothing to do.
      tfList[idx] := nil
      
      // We don't have to create it.
      shouldCreate = false
      
      // We can stop.
      break
    }
  }
  
  if shouldCreate {
    createResource(...)
  }
}

for _, tf := tfList {
  // It was not deleted from this list, so there is no related PR.
  if tf != nil {
    delete
  }
}
```

It still has that nested loop, but only once (and not twice as before) and we could remove two extra slices from the code (`oldTFs` and `newPRs`).

It may looks weird, but it can save a lot of cpu cycles and memory.

opudrovs replied:
@yitsushi added your optimized loops.

squaremo replied:
If you use a map, you don't need the inner loop:

```go
prSet := map[string]provider.PullRequest
for _, pr := range prs {
    prSet[fmt.Sprintf("%d", pr.Number)] = pr
}

for _, branchTf := range tfList {
    prLabel := branchTf.GetLabels()[LabelPRIDKey]
    if pr, ok := prSet[prLabel]; !ok {
        // no PR for this TF; delete it
        s.deleteTerraform(ctx, branchTf)
    } else {
        // don't need to create a Terraform for this PR, one exists
        delete(prMap, prLabel)
    }
}
for _, pr := range prMap {
    s.createTerraform(ctx, tf, source, pr.HeadBranch, fmt.Sprintf("%d", pr.Number), s.branchPollingInterval); err != nil {
}
```

I left out handling the errors from `create-` and `deleteTerraform`, even so it's a bit shorter I think.

The set does double-duty by keeping track of the set of Terraform objects that should exist and don't. There's a subtle assumption here, which is that at most one Terraform object will have any given PR number (otherwise, all but one will be deleted).

squaremo replied:
Two notes, though:

 1. It would be conventional to "upsert" all resources that are supposed to exist, since objects that already exist might nonetheless need their definitions updated. This simplifies things however you arrange loops. https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/controller/controllerutil has a func specifically for this.
 2. Having tests that demonstrate that it's correct would make it easier to change the particulars of the algorithm with confidence.

opudrovs replied:
1. What does upsert means in this context? Do you mean using
https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/controller/controllerutil#CreateOrUpdate

function instead of client.Create?

2.
> Having tests that demonstrate that it's correct would make it easier to change the particulars of the algorithm with confidence.

Do you mean specific tests for upserting objects? Are they different from tests for creating or deleting objects we already discussed?

squaremo replied:
By "upsert" I mean update if it needs updating (say if the original Terraform object had changed part of its spec), or create if it doesn't exist. This is really what the algorithm should do, I think, since you want to keep the branch Terraform objects up to date with respect to the original (and similarly the sources).

I mean tests that the algorithm does what it's supposed to: the equation is something like "given an original Terraform object and its source, for each pull request from the repo there should be a branch Terraform object the same as the original except with planOnly=true, and a source the same as the original source but referring to the pull request branch."

opudrovs replied:
Ah, you mean update branch TF objects not based on the PRs (I think no changes should be required for this) but based on the original TF object which might have changed.

So, you would like to add updating branch TFs in a loop too based on the original TF object (like all fields at once with modifying some of them after the update?))? If yes, what function from that package should be used for this?

squaremo reviewed (commented):
More comments while this is a draft. The only one I'd insist on (when this is ready for review) is the label name having a prefix particular to this controller; i.e., the prefix `infra.weave.works/`.

opudrovs commented on internal/informer/bbp/informer.go:
```diff
@@ -117,7 +118,7 @@ func (i *Informer) updateHandler(oldObj, newObj interface{}) {
 	}
 
-	if previous.Annotations[AnnotationKey] != AnnotationValue || current.Annotations[AnnotationKey] != AnnotationValue {
+	if previous.Labels[LabelKey] != LabelValue || current.Labels[LabelKey] != LabelValue {
```
It is possibly not related to the scope of the current PR.

@yitsushi should I change this line in the current PR after the create/delete objects logic was added?

yitsushi replied:
I don't think we have to check the previous object. I don't even know if we have to check **anything** on the previous object, or we need it at all. We care about the generated `ConfigMap` with the plan linked to the `current` object, and doesn't really matter what was the previous state, we don't use values from there.

I added that check there mostly as a filler to prevent go from crying about "unused variable" and I didn't want to write `_ = previous` while tried to capture the fact we have that available.

Summary: I think we can even delete everything about the `previous` object as I don't see any reasons to keep it. Something like this:
```diff
-func (i *Informer) updateHandler(oldObj, newObj interface{}) {
+func (i *Informer) updateHandler(_, newObj interface{}) {
// ...
-	previous, ok := oldObj.(*tfv1alpha2.Terraform)
-	if !ok {
-		i.log.Info("previous object is not a Terraform object", "object", oldObj)
-
-		return
-	}
// ...
-	if previous.Annotations[AnnotationKey] != AnnotationValue || current.Annotations[AnnotationKey] != AnnotationValue {
+	if current.Annotations[AnnotationKey] != AnnotationValue {

```

opudrovs commented on internal/server/polling/server.go:
```diff
@@ -97,9 +99,75 @@ func (s *Server) poll(ctx context.Context, resource types.NamespacedName, secret
 		s.log.Info("pull request", "pr", pr)
 	}
 
+	labels := map[string]string{
```
Yes, thanks, I have not checked the PR yet, it's still in draft.

opudrovs replied:
@squaremo so, about the labels. Do you prefer to change it to smth. like that:

```
tfList, err := s.listTerraformObjects(ctx, tf, map[string]string{
    bbp.LabelKey: bbp.LabelValue,
})
```
or create the selector

```
client.MatchingLabelsSelector{
    Selector: k8sLabels.Set(labels).AsSelector(),
}
```

somewhere outside the `listTerraformObjects` method because it does not change and just pass it then to the `listTerraformObjects` function? And if the latter, where are we supposed to store such pre-created things, which are not constants? Like just create a "getter" method building and returning it?

opudrovs replied:
For now changed it to

```
tfList, err := s.listTerraformObjects(ctx, tf, map[string]string{
    bbp.LabelKey: bbp.LabelValue,
})
```

opudrovs commented on internal/server/polling/terraform.go:
```diff
@@ -20,7 +22,28 @@ func (s *Server) getTerraform(ctx context.Context, ref client.ObjectKey) (*v1alp
+
+	tfObjects := []*infrav1.Terraform{}
+
+	for _, obj := range tfList.Items {
```
The problem without copying was that when creating the slice of pointers only the last object was added on all positions of the slice.

opudrovs replied:
Otherwise how can I convert a slice of TF objects (`Items`) to a slice of pointers to TF objects to return?

squaremo replied:
`tfList.Items` is exactly the value you want, if you don't care about copying.

opudrovs replied:
OK, thank you, got it!

```
result[i] = &tfList[i] // address of the entry
```

Here you mean `&tfList[i].Items`, correct?

squaremo replied:
> Here you mean `&tfList[i].Items`, correct?

Yip! Good spot

opudrovs commented on internal/server/polling/terraform.go:
```diff
@@ -20,7 +22,28 @@ func (s *Server) getTerraform(ctx context.Context, ref client.ObjectKey) (*v1alp
+	tfList := &infrav1.TerraformList{}
+
+	if err := s.clusterClient.List(ctx, tfList,
+		client.MatchingLabelsSelector{
```
> More comments while this is a draft. The only one I'd insist on (when this is ready for review) is the label name having a prefix particular to this controller; i.e., the prefix `infra.weave.works/`

So, you would like to change prefix for both keys?

```
LabelKey            = "terraform-controller/branch-based-planner"
LabelPRIDKey string = "terraform-controller/pr-id"
```

from `terraform-controller/` to `infra.weave.works/`?

yitsushi commented on internal/server/polling/server.go:
```diff
@@ -97,9 +100,75 @@ func (s *Server) poll(ctx context.Context, resource types.NamespacedName, secret
 		return fmt.Errorf("failed to list pull requests: %w", err)
 	}
 
+	// Log the PRs.
 	for _, pr := range prs {
 		s.log.Info("pull request", "pr", pr)
 	}
```
Don't forget to remove this.

opudrovs replied:
Still a draft PR. It's a leftover from your code. Log messages are there for convenience. Removed.

yitsushi commented on internal/server/polling/server.go:
```diff
@@ -97,9 +100,75 @@ func (s *Server) poll(ctx context.Context, resource types.NamespacedName, secret
+
+	// Delete TF objects without corresponding PRs.
+	for _, oldTF := range oldTFs {
+		err = s.deleteTerraform(ctx, oldTF)
```
I think we can move it up into the previous double-loop. Right now we have a quadratic double-loop that allocates and constantly resize a new slice, just to loop over all items and call delete.

Even setting the capacity of `oldTFs` to `len(tfList)` can save us a lot of resizing. Or simply just get rid of this slice.

yitsushi commented on internal/server/polling/server.go:
```diff
@@ -97,9 +100,75 @@ func (s *Server) poll(ctx context.Context, resource types.NamespacedName, secret
+	}
+
+	// Create Terraform objects for new PRs.
+	for _, pr := range newPRs {
```
Same as above.

yitsushi commented on internal/server/polling/terraform.go:
```diff
@@ -38,6 +62,119 @@ func (s *Server) getSource(ctx context.Context, tf *v1alpha2.Terraform) (*source
 }
 
+func (s *Server) createTerraform(ctx context.Context, originalTF *infrav1.Terraform, originalSource *sourcev1b2.GitRepository, branch string, prID string, interval time.Duration) error {
+	tf := originalTF.DeepCopy()
+
+	tf.ResourceVersion = ""
+
+	tfName := s.createObjectName(tf.ObjectMeta.Name, branch, prID)
+
+	tf.SetName(tfName)
+
+	tf.SetLabels(s.createLabels(tf.Labels, branch, prID))
+
+	msg := fmt.Sprintf("Terraform object %s in the namespace %s", tfName, tf.ObjectMeta.Namespace)
```
Can we group variable assignments and functions calls together? I like (and highly prefer) when the code has empty lines to break up logical sections, but not between every lines.

yitsushi commented on internal/server/polling/terraform.go:
```diff
@@ -38,6 +62,119 @@ func (s *Server) getSource(ctx context.Context, tf *v1alpha2.Terraform) (*source
+	return source, nil
+}
+
+func (s *Server) createObjectName(name string, branch string, prID string) string {
```
[can be ignored]
It's fine here, but later we may want to move these helper functions out of the struct as the Informer can use it, maybe it will not use it at all, but I like taking extra notes.

opudrovs commented:
>I think as a first step, we can merge the two nested loops (delete tfs and create tfs).

I've already optimized the first quadratic loop, thinking about what t do with the second one. Will look into your code.

opudrovs commented:
@squaremo @yitsushi 

What is preferred:

to use

```
SetName
GetName
SetNamespace
GetNamespace
```

and similar getters/setters for `Labels` and `Annotations`

vs

```
Name
Namespace
Labels
Annotations
```
to get and set values of these fields?

Are both ways functionally the same and does using getters/setters explicitly just prevent us from setting (`obj.GetName() = "blablabla"`) or assigning/returning (like `x := obj.SetName("blablabla")`) values accidentally, which in the examples will result in errors?

opudrovs commented:
In addition, is it ever recommended to use `obj.ObjectMeta.Name` instead of `obj.Name` in such cases?

Or should `ObjectMeta` be used only when initializing structures like

```
ObjectMeta: metav1.ObjectMeta{
    Name:      "blablaname",
    Namespace: "blablanamespace",
},
```
not when assigning values to individual fields like `obj.ObjectMeta.Name = "blablabla"`?

squaremo commented on cmd/branch-based-planner/flags.go:
```diff
@@ -31,10 +32,18 @@ func parseFlags() *applicationOptions {
 
+	flag.DurationVar(&opts.branchPollingInterval,
+		"branch-polling-interval", 0,
+		"Wait between two request to the same Terraform object.")
```
```suggestion
		"Interval to use for PR branch sources (default is to use the value of --polling-interval)")
```

opudrovs commented:
@squaremo @yitsushi so, I've added the `CreateOrUpdate` functions.

The main question:

What to do with setting `ResourceVersion`?

```
tf.SetResourceVersion("")
...
source.SetResourceVersion("")
```

If I don't add this, there is an error when creating objects:

```
{"level":"error","ts":"2023-06-28T20:56:25.758+0200","logger":"polling-server","msg":"failed to reconcile Terraform object","version":"","sha":"","error":"reconcile failed for Terraform object helloworld-tf-test-4-4 in the namespace default: resourceVersion should not be set on objects to be created"}
```

If I add this, the branch TF objects are created but are not updated after applying the modified YAML for the original object.

After updating the original YAML, I see a message that the object were updated but the fields I changed in the original TF are not updated:

```
{"level":"info","ts":"2023-06-28T21:13:22.576+0200","logger":"polling-server","msg":"Source helloworld-test-4-4 in the namespace default successfully reconciled","version":"","sha":"","operation":"unchanged"}

{"level":"info","ts":"2023-06-28T21:13:22.592+0200","logger":"polling-server","msg":"Terraform object helloworld-tf-test-4-4 in the namespace default successfully reconciled","version":"","sha":"","operation":"updated"}
```

If I got it right, the actual `ResourceVersion` is required for updates.

I tried to use the following check from the method docs

```
obj.ObjectMeta.CreationTimestamp.IsZero()
```

when setting the `ResourceVersion` to `""`,but it always equals false, outside or within the mutate functions.

opudrovs commented:
@squaremo @yitsushi   Additional questions:

2) Should these methods be called `createOrUpdateTerraform` and `createOrUpdateSource` (instead of create) or `reconcileTerraform` and `reconcileSource` (the latter might help to avoid confusion)? Because the messages in the example are speaking of reconciliation and our server method is called `reconcile`.

3) Is returning the source from `createOrUpdateSource` method OK? I am not sure it will be updated by that method (the original `CreateOrUpdate`), but we only need it to get its name and namespace for the TF spec.

4) I changed the verbs in the log messages on failure or successful creation/update from `create` or `update` to `reconcile`, similar to the example. Is it OK?

opudrovs commented:
I've added running tests in the `internal` directory and subdirectories to the `test` target. But as far as I can see, the `test` target is not run in any workflows, it can be run only locally. And I am getting a timeout for controller tests for some reason when running this target with added internal tests locally. Maybe this should be tacked in a separate issue.

squaremo reviewed (dismissed):
:100: for including tests, asking questions about the best way to do things, and otherwise for responding to the extended commentary :-)

May I request a couple of things before this is merged:
 - fix the `.PHONY` line (I've commented at the appropriate line)
 - update the commit summary and include a prose description of the changes in the commit message*.

(*here is some great advice on what to put in commit messages: https://cbea.ms/git-commit/#why-not-how)
